inc/options/general.php

```diff
@@ -661,13 +661,103 @@ class="w3tc-ignore-change" type="text"
 			<?php Util_Ui::postbox_footer(); ?>
 		<?php endif; ?>
 
+		<?php Util_Ui::postbox_header( __( 'Google Page Speed', 'w3-total-cache' ), '', 'google_page_speed' ); ?>
 		<?php
-		foreach ( $custom_areas as $area ) {
-			do_action( "w3tc_settings_general_boxarea_{$area['id']}" );
+		$access_token_json = ( ! empty( $this->_config->get_string( 'widget.pagespeed.access_token' ) ) ? $this->_config->get_string( 'widget.pagespeed.access_token' ) : '' );
+		$w3_pagespeed      = new PageSpeed_Api( $access_token_json );
+
+		$site_id    = Util_Http::generate_site_id();
+		$return_url = admin_url( 'admin.php?page=w3tc_general' );
+		$w3key      = ! empty( $this->_config->get_string( 'widget.pagespeed.w3key' ) ) ? $this->_config->get_string( 'widget.pagespeed.w3key' ) : '';
+		$auth_url   = $w3_pagespeed->client->createAuthUrl();
+
+		// This may need a nonce or equivilant security measure?
```

Comment:
Are you asking me or telling me?

Reply:
I recall during my research seeing that a nonce or equivalent could be passed to the authorization/oauth2 request to prevent CSRF but for this it's probably unnecessary